x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeComputeHandler.java

```diff
@@ -98,15 +98,25 @@ void startComputeOnDataNodes(
         Runnable runOnTaskFailure,
         ActionListener<ComputeResponse> outListener
     ) {
-        final boolean allowPartialResults = configuration.allowPartialResults();
-        DataNodeRequestSender sender = new DataNodeRequestSender(transportService, esqlExecutor, parentTask, allowPartialResults) {
+        new DataNodeRequestSender(
+            transportService,
+            esqlExecutor,
+            parentTask,
+            configuration.allowPartialResults(),
+            configuration.pragmas().maxConcurrentNodePerCluster()
+        ) {
             @Override
             protected void sendRequest(
                 DiscoveryNode node,
                 List<ShardId> shardIds,
                 Map<Index, AliasFilter> aliasFilters,
                 NodeListener nodeListener
             ) {
+                if (exchangeSource.isCompleted()) {
+                    nodeListener.onResponse(new DataNodeComputeResponse(List.of(), Map.of()));
+                    return;
+                }
```

Comment:
This part prevents us from sending a query to remaining data nodes if we collected enough results

Reply:
It looks like we won't send more requests because we do:
```
                if (skipRemaining) {
                    DataNodeRequestSender.this.skipRemaining = true;
                }
```

So we'll only count the number of shards we skip and that's it. I think.

Reply:
> It looks like we won't send more requests because we do:

Correct, this was added today in: https://github.com/elastic/elasticsearch/pull/122850/commits/a5084ab124b1ec79cb0e24c9d84ac30f7f844dc9

> So we'll only count the number of shards we skip and that's it. I think.

The total skipped count consists of ones we skipped already (`skippedShards`) and count of shards we have not processed (remaining shards in `pendingShardIds`), please see https://github.com/elastic/elasticsearch/blob/50da08791a91983996348b3c544ef63421b4cb4e/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java#L96

Reply:
Since these computations should not be expensive, I wonder if we should skip only here, not shortcutting in other places. The reason is that we might need to be more careful not to shortcut in other places when allow_partial_results=true.